src/stream/src/common/table/state_table.rs

```diff
@@ -883,11 +884,7 @@ where
                     let mut streams = vec![];
                     for vnode in self.vnodes().iter_vnodes() {
                         let stream = self
-                            .iter_key_and_val_with_pk_range(
-                                &range,
-                                vnode,
-                                PrefetchOptions::default(),
-                            )
+                            .iter_kv_with_pk_range(&range, vnode, PrefetchOptions::default())
```

Comment:
I will fix this in a separate PR, I think it should iter on pk here, rather than raw key.

Otherwise https://github.com/risingwavelabs/risingwave/issues/11869 is encountered.

Reply:
I guess this is a bug prior to this PR? I thought the intention is to merge_sort with vnode included. This is another bug caused by the confusing API lol

Reply:
> I will fix this in a separate PR, I think it should iter on pk here, rather than raw key.

Couldn't quite get it here. I think this by splitting vnode and pk bytes, the `merge_sort` can be corrected, and then #11869 will no longer be an issue.

Reply:
> I thought the intention is to merge_sort with vnode included.

No, `merge_sort` is intended to sort rows from different vnodes by their pk without vnode prefix.

Reply:
> I guess this is a bug prior to this PR? I thought the intention is to merge_sort with vnode included. This is another bug caused by the confusing API lol

Yes it is prior to this PR, out of scope and can be fixed separately.

Reply:
> Couldn't quite get it here. I think by splitting vnode and pk bytes in this PR, the merge_sort can be corrected, and then https://github.com/risingwavelabs/risingwave/issues/11869 will no longer be an issue.

I mean here we still use `iter_kv_with_pk_range`. Underlying that is:
```
pub type StateStoreIterItem = (FullKey<Bytes>, Bytes);
```

The `FullKey` includes `vnode` as well. So if you look later in the code, that still uses `deserialize_pk_with_vnode` to deserialize, which is incorrect.

After this PR, `iter_row_and_key_with_pk_range` will return :
```
pub type RowStreamWithKey<'a, S: StateStore, SD: ValueRowSerde + 'a> =
    impl Stream<Item = StreamExecutorResult<((VirtualNode, Bytes), OwnedRow)>> + 'a;
```

We want the new iter to be something like `iter_pk_and_row_with_pk_range`.

Which is just a simple wrapper around `iter_row_and_key_with_pk_range`. This will drop the vnode, part, and ensure when `merge_sort` happens, it will only sort by `pk`.

Reply:
Then we're at the same page here. While from my perspective, it doesn't need to be made too complex.

Since we provide `vnode` and `pk_range` separately when calling `iter_kv_with_pk_range`, it should return `vnode` and `pk` separately, I guess that why this PR exists. And once we have separated `vnode` and `pk`, we need to rethink `merge_sort`. AFAIK from its very first use in streaming executors, it's supposed to sort rows from different iter streams by `pk` instead of `vnode + pk`.

So the two issues are highly related and `merge_sort` can be fixed in this PR by just use `peeked = stream.next().await?.0.1`, or, for better readability we may consider put `vnode` and `pk` in a struct instead of anonymous tuple.

Reply:
My point is that we don't need introduce a new state table method or wrap the streams before calling `merge_sort` to fix it.

Reply:
>  for better readability we may consider put vnode and pk in a struct instead of anonymous tuple.


> My point is that we don't need introduce a new state table method or wrap the streams before calling merge_sort to fix it.

+1 for both

Reply:
Oh I realized that we can't get `1` from the key in `merge_sort`. Just ignore me😇

Reply:
Another reasonable solution may be wrapping `vnode` and `pk` in  a struct and impl `Ord` for it to make it sort by `pk`. IMO it's reasonable because `pk` is the key the caller want to care about, state table should hide vnode concept (for almost all executors, the order of vnodes doesn't matter).

> Some components will still want to iter with vnode, e.g. arrangement backfill.

For some of such executors that need to care about vnode, they may simply chain all streams when calling `iter_vnodes` and creating iteration streams.

Reply:
Anyway, just some concern on elegance, no strong opinion. All solutions mentioned in previous comments LGTM🤪